---
editor_options: 
  markdown: 
    wrap: 72
---

# Filtering

### Post-pipeline (selection menu)

| Type | Description |
|:-------------------|:---------------------------------------------------|
| Annotation | the annotated VCF produced by the Sarek pipeline |
| Region | filtered by QUAL \>= 100, Genotype Quality \>= 80, Depth \>= 10, Genome-in-a-bottle all difficult regions |
| Population | Max AF from 1000 genomes or gnomAD \< 0.01 |
| Mutation | significant mutations |
| ML Driver Genes | Any mutations to selected driver genes pulled from the Annotation VCF file |

### Pipeline (FILTER Column)

| Value | Description |
|:----------------------|:------------------------------------------------|
| FAIL | Fail the site if all alleles fail but for different reasons. |
| base_qual | alt median base quality |
| clustered_events | Clustered events observed in the tumor |
| contamination | contamination |
| duplicate | evidence for alt allele is overrepresented by apparent duplicates |
| fragment | abs(ref - alt) median fragment length |
| germline | Evidence indicates this site is germline, not somatic |
| haplotype | Variant near filtered variant on same haplotype. |
| low_allele_frac | Allele fraction is below specified threshold |
| map_qual | ref - alt median mapping quality |
| multiallelic | Site filtered because too many alt alleles pass tumor LOD |
| n_ratio | Ratio of N to alt exceeds specified ratio |
| normal_artifact | artifact_in_normal |
| orientation | orientation bias detected by the orientation bias mixture model |
| panel_of_normals | Blacklisted site in panel of normals |
| position | median distance of alt variants from end of reads |
| possible_numt | Allele depth is below expected coverage of NuMT in autosome |
| slippage | Site filtered due to contraction of short tandem repeat region |
| strand_bias | Evidence for alt allele comes from one read direction only |
| strict_strand | Evidence for alt allele is not represented in both directions |
| weak_evidence | Mutation does not meet likelihood threshold |

# FORMAT

### Haplotypecaller
| Label | Description |
|:------|:------------|
|AD | Allelic depths for the ref and alt alleles in the order listed|
|AF|Allele fractions of alternate alleles in the tumor|
|DP|Approximate read depth (reads with MQ 255 or with bad mates are filtered)|
|F1R2|Count of reads in F1R2 pair orientation supporting each allele|
|F2R1|Count of reads in F2R1 pair orientation supporting each allele|
|FAD|Count of fragments supporting each allele.|
|GQ|Genotype Quality|
|GT|Genotype|
|PGT|Physical phasing haplotype information, describing how the alternate alleles are phased in relation to one another; will always be heterozygous and is not intended to describe called alleles|
|PID|Physical phasing ID information, where each unique ID within a given sample (but not across samples) connects records within a phasing group|
|PL|Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification|
|PS|Phasing set (typically the position of the first variant in the set)|
|SB|Per-sample component statistics which comprise the Fisher's Exact Test to detect strand bias.|

### Deepvariant

### Mutect2

### Strelka

### Manta

# INFO

### Haplotypecaller
| Label | Description |
|:------|:------------|
|AC|Allele count in genotypes, for each ALT allele, in the same order as listed|
|AF|Allele Frequency, for each ALT allele, in the same order as listed|
|AN|Total number of alleles in called genotypes|
|BaseQRankSum|Z-score from Wilcoxon rank sum test of Alt Vs. Ref base qualities|
|CNN_1D|Log odds of being a true variant versus being false under the trained 1D Convolutional Neural Network| 
|DB|dbSNP Membership|
|DP|Approximate read depth; some reads may have been filtered|
|ExcessHet|Phred-scaled p-value for exact test of excess heterozygosity|
|FS|Phred-scaled p-value using Fisher's exact test to detect strand bias|
|InbreedingCoeff|Inbreeding coefficient as estimated from the genotype likelihoods per-sample when compared against the Hardy-Weinberg expectation|
|MLEAC|Maximum likelihood expectation (MLE) for the allele counts (not necessarily the same as the AC), for each ALT allele, in the same order as listed|
|MLEAF|Maximum likelihood expectation (MLE) for the allele frequency (not necessarily the same as the AF), for each ALT allele, in the same order as listed|
|MQ|RMS Mapping Quality|
|MQRankSum|Z-score From Wilcoxon rank sum test of Alt vs. Ref read mapping qualities|
|QD|Variant Confidence/Quality by Depth|
|ReadPosRankSum|Z-score from Wilcoxon rank sum test of Alt vs. Ref read position bias|
|SOR|Symmetric Odds Ratio of 2x2 contingency table to detect strand bias|

| Type | Description |
|:-----|:------------|
|AS_FilterStatus|Filter status for each allele, as assessed by ApplyVQSR. Note that the VCF filter field will reflect the most lenient/sensitive status across all alleles.| 
|AS_SB_TABLE|Allele-specific forward/reverse read counts for strand bias tests. Includes the reference and alleles separated by \|.|
|AS_UNIQ_ALT_READ_COUNT|Number of reads with unique start and mate end positions for each alt at a variant site|
|CONTQ|Phred-scaled qualities that alt allele are not due to contamination|
|DP|Approximate read depth; some reads may have been filtered|
|ECNT|Number of events in this haplotype|
|GERMQ|Phred-scaled quality that alt alleles are not germline variants|
|MBQ|median base quality by allele|
|MFRL|median fragment length by allele|
|MMQ|median mapping quality by allele|
|MPOS|median distance from end of read|
|NALOD|Negative log 10 odds of artifact in normal with same allele fraction as tumor|
|NCount|Count of N bases in the pileup|
|NLOD|Normal log 10 likelihood ratio of diploid het or hom alt genotypes|
|OCM|Number of alt reads whose original alignment doesn't match the current contig.|
|PON|site found in panel of normals|
|POPAF|negative log 10 population allele frequencies of alt alleles|
|ROQ|Phred-scaled qualities that alt allele are not due to read orientation artifact|
|RPA|Number of times tandem repeat unit is repeated, for each allele (including reference)|
|RU|Tandem repeat unit (bases)|
|SEQQ|Phred-scaled quality that alt alleles are not sequencing errors|
|STR|Variant is a short tandem repeat|
|STRANDQ|Phred-scaled quality of strand bias artifact|
|STRQ|Phred-scaled quality that alt alleles in STRs are not polymerase slippage errors|
|TLOD|Log 10 likelihood ratio score of variant existing versus not existing|
|LOF|Predicted loss of function effects for this variant. Format: Gene_Name \| Gene_ID \| Number_of_transcripts_in_gene \| Percent_of_transcripts_affected|
|NMD|Predicted nonsense mediated decay effects for this variant. Format: Gene_Name \| Gene_ID \| Number_of_transcripts_in_gene \| Percent_of_transcripts_affected|

### Deepvariant

### Mutect2

### Strelka

### Manta

# ANNOTATIONS
For more detail see: [PMC7709417](https://pmc.ncbi.nlm.nih.gov/articles/PMC7709417/)

| Type | Description |
|:-------------------|:---------------------------------------------------|
|CADD_phred |CADD_phred from dbNSFP file.| 
|DANN_score |DANN_score from dbNSFP file.| 
|ExAC |ExAC from dbNSFP file.| 
|FATHMM_pred |FATHMM_pred from dbNSFP file.| 
|Interpro_domain |Interpro_domain from dbNSFP file.|
|LRT_pred |LRT_pred from dbNSFP file.| 
|MetaSVM_pred |MetaSVM_pred from dbNSFP file.| 
|MutationTaster_pred |MutationTaster_pred from dbNSFP file.|
|PROVEAN_pred |PROVEAN_pred from dbNSFP file.| 
|Polyphen2_HDIV_pred |Polyphen2_HDIV_pred from dbNSFP file.| 
|Polyphen2_HVAR_pred |Polyphen2_HVAR_pred from dbNSFP file.| 
|PrimateAI_pred |PrimateAI_pred from dbNSFP file.| 
|REVEL_score |REVEL_score from dbNSFP file.|
|SIFT_pred |SIFT_pred from dbNSFP file.| 
|dbSNP |dbSNP from dbNSFP file.|
